File: ios/PassportMRZReadingObjC/PassportMRZReadingObjC/LabelResultView.m
//
//  LabelResultView.m
//  PassportMRZReadingObjC
//
//  Copyright © 2021 dynamsoft. All rights reserved.
//

#import "LabelResultView.h"

@interface LabelResultView ()<UITableViewDelegate, UITableViewDataSource>

@end

@implementation LabelResultView{
    UITableView* resTableview;
    UIImageView* resImageview;
    UILabel* tipsImage;
    UILabel* tipsTry;
    UILabel* titleLabel;
    UIView* backView;
}

- (void)viewDidLoad{
    [super viewDidLoad];
    
    CGFloat h = self.view.bounds.size.height;
    CGFloat w = self.view.bounds.size.width;
    CGFloat tabH = self.tabBarController.tabBar.bounds.size.height + 64;
    resImageview = [[UIImageView alloc] initWithFrame:CGRectMake(w / 6, tabH, w * 2 / 3, h / 3)];
    [resImageview setImage:_recogImg];
    if (_recogImg.imageOrientation == UIImageOrientationRight) {
        CGAffineTransform transform = CGAffineTransformIdentity;
        resImageview.transform = CGAffineTransformRotate(transform, -M_PI_2);
        resImageview.frame = CGRectMake(w / 6, tabH, w * 2 / 3, h / 3);
    }
    
    titleLabel = [[UILabel alloc] initWithFrame:CGRectMake(0, tabH - 40, w * 2 / 3, 40)];
    titleLabel.text = @"PassportMRZReading";
    titleLabel.textColor = UIColor.blackColor;
    [titleLabel setFont:[UIFont boldSystemFontOfSize:20]];
    
    backView = [[UIView alloc] initWithFrame:CGRectMake(w - 60, tabH - 40, 40, 40)];
    UIImage* bg = [UIImage imageNamed:@"close"];
    [backView layer].contents = (id)bg.CGImage;
    UITapGestureRecognizer *gr = [[UITapGestureRecognizer alloc]initWithTarget:self action:@selector(back)];
    [backView addGestureRecognizer:gr];
    
    tipsImage = [[UILabel alloc] initWithFrame:CGRectMake(0, h / 3 + tabH, w, 30)];
    tipsImage.textColor = UIColor.blackColor;
    tipsImage.text = @"Capture passport Image";
    tipsImage.textAlignment = NSTextAlignmentCenter;
    tipsTry = [[UILabel alloc] initWithFrame:CGRectMake(self.view.center.x - 40, h / 3 + tabH + 30, 80, 30)];
    tipsTry.textColor = UIColor.blackColor;
    NSDictionary *attribtDic = @{NSUnderlineStyleAttributeName : [NSNumber numberWithInteger:NSUnderlineStyleSingle], NSUnderlineColorAttributeName : [UIColor blackColor]};
    NSMutableAttributedString *attribtStr = [[NSMutableAttributedString alloc]initWithString:@"Try it out" attributes:attribtDic];
    tipsTry.attributedText = attribtStr;
    UITapGestureRecognizer *gestureRecognizer = [[UITapGestureRecognizer alloc]initWithTarget:self action:@selector(Click)];
    [tipsTry addGestureRecognizer:gestureRecognizer];
    tipsTry.userInteractionEnabled = YES;
    
    resTableview = [[UITableView alloc] initWithFrame:CGRectMake(5, h / 3 + tabH + 60, w - 10, h / 2 - 64) style:UITableViewStylePlain];
    resTableview.delegate = self;
    resTableview.dataSource = self;
    resTableview.separatorColor = UIColor.grayColor;
    resTableview.userInteractionEnabled = YES;
    [self.view addSubview:titleLabel];
    [self.view addSubview:backView];
    [self.view addSubview:resImageview];
    [self.view addSubview:tipsImage];
    [self.view addSubview:resTableview];
}

- (void)back{
    [self.presentingViewController dismissViewControllerAnimated:YES completion:nil];
}

- (void)Click{
    //try it out
}

- (BOOL)prefersStatusBarHidden{
    return YES;
}

- (void)addCell:(UITableViewCell*)cell txt:(NSString*)txt{
    CGSize size = CGSizeMake(100, 40);
    CGFloat tablex = resTableview.frame.origin.x + resTableview.frame.size.width;
    UILabel *lb = [[UILabel alloc] initWithFrame:CGRectMake(tablex - size.width, 40 - size.height / 2, 100, 40)];
    lb.textColor = UIColor.grayColor;
    lb.text = txt;
    lb.font = [UIFont boldSystemFontOfSize:18];
    lb.textAlignment = NSTextAlignmentCenter;
    CGSize sizeNew = [lb.text sizeWithAttributes:@{NSFontAttributeName:[UIFont boldSystemFontOfSize:18]}];
    lb.frame = CGRectMake(tablex - sizeNew.width - 15, 40 - sizeNew.height / 2, sizeNew.width, sizeNew.height);
    [cell.contentView addSubview:lb];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath{
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"mrz"];
    for (UIView* v  in cell.contentView.subviews) {
        [v removeFromSuperview];
    }
    if (!cell) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:@"mrz"];
    }
    NSString* title = @"";
    switch (indexPath.row) {
        case 0:
            title = @"SURNAME";
            [self addCell:cell txt:_contentArr[1]];
            break;
        case 1:
            title = @"GIVEN NAMES";
            [self addCell:cell txt:_contentArr[2]];
            break;
        case 2:
            title = @"NATIONALITY";
            [self addCell:cell txt:_contentArr[0]];
            break;
        case 3:
            title = @"SEX/GENDER";
            [self addCell:cell txt:_contentArr[6]];
            break;
        case 4:
            title = @"DATE OF BIRTH";
            [self addCell:cell txt:[self getDate:_contentArr[5]]];
            break;
        case 5:
            title = @"ISSUING COUNTRY";
            [self addCell:cell txt:_contentArr[4]];
            break;
        case 6:
            title = @"PASSPORT NUMBER";
            [self addCell:cell txt:_contentArr[3]];
            break;
        case 7:
            title = @"PASSPORT EXPIRATION";
            [self addCell:cell txt:[self getDate:_contentArr[7]]];
            break;
        default:
            break;
    }
    cell.textLabel.text = title;
    cell.textLabel.font = [UIFont italicSystemFontOfSize:16];
    return cell;
}

- (NSString*)getDate:(NSString*)str{
    int mm = [[str substringWithRange:NSMakeRange(2, 2)] intValue];
    NSArray* vals = [NSArray arrayWithObjects:@"Jan",@"Feb",@"Mar",@"Apr",@"May",@"Jun",@"Jul",@"Aug",@"Sep",@"Oct",@"Nov",@"Dec", nil];
    NSString* year = [str substringWithRange:NSMakeRange(0, 2)];
    NSString* day = [str substringWithRange:NSMakeRange(4, 2)];
    NSString* mo = vals[mm - 1];
    NSString* ret = [[day stringByAppendingFormat:@" %@",mo] stringByAppendingFormat:@" %@",year];
    return ret;
}

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
    return 80;
}

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView{
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section{
    return 8;
}

- (void)viewWillAppear:(BOOL)animated{
    [super viewWillAppear:animated];
}

- (void)viewWillDisappear:(BOOL)animated{
    [super viewWillDisappear:animated];
}

- (void)viewWillTransitionToSize:(CGSize)size withTransitionCoordinator:(id<UIViewControllerTransitionCoordinator>)coordinator{
    CGFloat w = size.width;
    CGFloat h = size.height;
    CGFloat tabH = self.tabBarController.tabBar.bounds.size.height + 64;
    resImageview.frame = CGRectMake(w / 6, tabH, w * 2 / 3, h / 3);
    if (_recogImg.imageOrientation == UIImageOrientationRight) {
        CGAffineTransform transform = CGAffineTransformIdentity;
        resImageview.transform = CGAffineTransformRotate(transform, -M_PI_2);
        resImageview.frame = CGRectMake(w / 6, tabH, w * 2 / 3, h / 3);
    }
    titleLabel.frame = CGRectMake(0, tabH - 40, w * 2 / 3, 40);
    backView.frame = CGRectMake(w - 60, tabH - 40, 40, 40);
    tipsImage.frame = CGRectMake(0, h / 3 + tabH, w, 30);
    resTableview.frame = CGRectMake(5, h / 3 + tabH + 60, w - 10, h / 2 - 64);
    [resTableview reloadData];
}

@end

File: ios/PassportMRZReadingSwift/PassportMRZReadingSwift/LabelResultView.swift
//
//  LabelResultView.swift
//  PassportMRZReadingSwift
//
//  Copyright © 2021 Dynamsoft. All rights reserved.
//

import UIKit

class LabelResultView: UIViewController, UITableViewDelegate, UITableViewDataSource {
    var recogImg:UIImage!
    var contentArr:[String] = []
    
    var resTableview:UITableView!
    var resImageview:UIImageView!
    var tipsImage:UILabel!
    var titleLabel:UILabel!
    var backView:UIView!
    
    let w = UIScreen.main.bounds.size.width
    let h = UIScreen.main.bounds.size.height
    let tabH = UIApplication.shared.statusBarFrame.size.height + 64
    override func viewDidLoad() {
        super.viewDidLoad()
        resImageview = UIImageView(frame: CGRect(x:w / 6, y: tabH, width: w * 2 / 3, height: h/3))
        resImageview.image = recogImg
        if (recogImg.imageOrientation == .right) {
            let transform = CGAffineTransform.identity
            resImageview.transform = transform.rotated(by: CGFloat(-Double.pi / 2))
            resImageview.frame = CGRect(x:w / 6, y: tabH, width: w * 2 / 3, height: h/3)
        }
        
        titleLabel = UILabel(frame: CGRect(x:0, y: tabH - 40, width: w * 2 / 3, height: 40))
        titleLabel.text = "PassportMRZReading"
        titleLabel.textColor = UIColor.black
        titleLabel.font = .boldSystemFont(ofSize: 20)
        backView = UIView(frame: CGRect(x:w - 60, y: tabH - 40, width: 40, height: 40))
        let bg = UIImage(named: "close")
        backView.layer.contents = bg?.cgImage
        let gr = UITapGestureRecognizer(target: self, action: #selector(back))
        backView.addGestureRecognizer(gr)
        tipsImage = UILabel(frame: CGRect(x:0, y: h / 3 + tabH, width: w, height: 30))
        tipsImage.textColor = UIColor.black
        tipsImage.text = "Capture passport Image"
        tipsImage.textAlignment = .center
        resTableview = UITableView(frame: CGRect(x:5, y: h / 3 + tabH + 60, width: w - 10, height: h / 2 - 64), style: .plain)
        resTableview.delegate = self
        resTableview.dataSource = self
        resTableview.separatorColor = UIColor.gray
        resTableview.isUserInteractionEnabled = true
        
        self.view.addSubview(titleLabel)
        self.view.addSubview(backView)
        self.view.addSubview(resImageview)
        self.view.addSubview(tipsImage)
        self.view.addSubview(resTableview)
    }
    
    @objc func back(){
        self.presentingViewController?.dismiss(animated: true, completion: nil)
    }
    
    func addCell(cell:UITableViewCell, txt:String){
        let size = CGSize(width: 100, height: 40)
        let tablex = resTableview.frame.origin.x + resTableview.frame.size.width
        let lb = UILabel(frame: CGRect(x:tablex - size.width, y: 40 - size.height / 2, width: 100, height: 40))
        lb.textColor = UIColor.gray
        lb.text = txt
        lb.font = .boldSystemFont(ofSize: 18)
        lb.textAlignment = .center
        let str:NSString = lb.text! as NSString
        let sizeNew = str.size(withAttributes: [NSAttributedStringKey.font:UIFont.boldSystemFont(ofSize: 18)])
        lb.frame = CGRect(x:tablex - sizeNew.width - 15, y:40 - sizeNew.height / 2, width: sizeNew.width, height:sizeNew.height)
        cell.contentView.addSubview(lb)
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        var cell = tableView.dequeueReusableCell(withIdentifier: "mrz")
        if cell != nil {
            for v in cell!.contentView.subviews {
                v.removeFromSuperview()
            }
        }
        
        if cell == nil {
            cell = UITableViewCell(style: .default, reuseIdentifier: "mrz")
        }
        var title = ""
        switch indexPath.row {
        case 0:
            title = "SURNAME"
            self.addCell(cell: cell!, txt: contentArr[1])
        case 1:
            title = "GIVEN NAMES"
            self.addCell(cell: cell!, txt: contentArr[2])
        case 2:
            title = "NATIONALITY"
            self.addCell(cell: cell!, txt: contentArr[0])
        case 3:
            title = "SEX/GENDER"
            self.addCell(cell: cell!, txt: contentArr[6])
        case 4:
            title = "DATE OF BIRTH"
            self.addCell(cell: cell!, txt: self.getDate(str: contentArr[5]))
        case 5:
            title = "ISSUING COUNTRY"
            self.addCell(cell: cell!, txt: contentArr[4])
        case 6:
            title = "PASSPORT NUMBER"
            self.addCell(cell: cell!, txt: contentArr[3])
        case 7:
            title = "PASSPORT EXPIRATION"
            self.addCell(cell: cell!, txt: self.getDate(str: contentArr[7]))
        default:
            title = ""
        }
        cell?.textLabel?.text = title
        cell?.textLabel?.font = .italicSystemFont(ofSize: 16)
        return cell!
    }
    
    func getDate(str: String) -> String {
        let mm:Int = Int(str[2..<4])!
        let vals = ["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"]
        let year = str[0..<2]
        let day = str[4..<6]
        let mo = vals[mm - 1]
        let ret = day.appending(" \(String(describing: mo)) \(year)")
        return ret
    }
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        return 80
    }
    
    func numberOfSections(in tableView: UITableView) -> Int {
        return 1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 7
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
    }
    
    override var prefersStatusBarHidden: Bool{
        return true
    }
    
    override func viewWillTransition(to size: CGSize, with coordinator: UIViewControllerTransitionCoordinator) {
        let w = size.width
        let h = size.height
        resImageview.frame = CGRect(x:w / 6, y: tabH, width: w * 2 / 3, height: h/3)
        if (recogImg.imageOrientation == .right) {
            let transform = CGAffineTransform.identity
            resImageview.transform = transform.rotated(by: CGFloat(-Double.pi / 2))
            resImageview.frame = CGRect(x:w / 6, y: tabH, width: w * 2 / 3, height: h/3)
        }
        titleLabel.frame = CGRect(x:0, y: tabH - 40, width: w * 2 / 3, height: 40)
        backView.frame = CGRect(x:w - 60, y: tabH - 40, width: 40, height: 40)
        tipsImage.frame = CGRect(x:0, y: h / 3 + tabH, width: w, height: 30)
        resTableview.frame = CGRect(x:5, y: h / 3 + tabH + 60, width: w - 10, height: h / 2 - 64)
        resTableview.reloadData()
    }
}
